libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp

```diff
@@ -0,0 +1,356 @@
+#include <AP_RCTelemetry/AP_CRSF_Telem.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+
+/*
+ * CRSF protocol
+ *
+ * CRSF protocol uses a single wire half duplex uart connection.
+ * The master sends one frame every 4ms and the slave replies between two frames from the master.
+ *
+ * 420000 baud
+ * not inverted
+ * 8 Bit
+ * 1 Stop bit
+ * Big endian
+ * 416666 bit/s = 46667 byte/s (including stop bit) = 21.43us per byte
+ * Max frame size is 64 bytes
+ * A 64 byte frame plus 1 sync byte can be transmitted in 1393 microseconds.
+ *
+ * CRSF_TIME_NEEDED_PER_FRAME_US is set conservatively at 1500 microseconds
+ *
+ * Every frame has the structure:
+ * <Device address><Frame length><Type><Payload><CRC>
+ *
+ * Device address: (uint8_t)
+ * Frame length:   length in  bytes including Type (uint8_t)
+ * Type:           (uint8_t)
+ * CRC:            (uint8_t)
+ *
+ */
+
+extern const AP_HAL::HAL& hal;
+
+// #define CRSF_DEBUG
+#ifdef CRSF_DEBUG
+# define debug(fmt, args...)	hal.console->printf("CRSF: " fmt "\n", ##args)
+static const char* get_frame_type(uint8_t byte)
+{
+    switch(byte) {
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_GPS:
+        return "GPS";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_BATTERY_SENSOR:
+        return "BATTERY";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_HEARTBEAT:
+        return "HEARTBEAT";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_VTX:
+        return "VTX";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_VTX_TELEM:
+        return "VTX_TELEM";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_PARAM_DEVICE_PING:
+        return "PING";
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_COMMAND:
+        return "COMMAND";
+    default:
+        return "UNKNOWN";
+    }
+}
+#else
+# define debug(fmt, args...)	do {} while(0)
+#endif
+
+#define CRSF_MAX_FRAME_TIME_US      1100U // 700us + 400us for potential ad-hoc request
+#define CRSF_INTER_FRAME_TIME_US_150HZ    6667U // At fastest, frames are sent by the transmitter every 6.667 ms, 150 Hz
+#define CRSF_INTER_FRAME_TIME_US_50HZ    20000U // At slowest, frames are sent by the transmitter every 20ms, 50 Hz
+#define CSRF_HEADER_LEN     2
+
+#define CRSF_DIGITAL_CHANNEL_MIN 172
+#define CRSF_DIGITAL_CHANNEL_MAX 1811
+
+
+AP_RCProtocol_CRSF* AP_RCProtocol_CRSF::_singleton;
+
+AP_RCProtocol_CRSF::AP_RCProtocol_CRSF(AP_RCProtocol &_frontend) : AP_RCProtocol_Backend(_frontend)
+{
+#if !APM_BUILD_TYPE(APM_BUILD_UNKNOWN)
+    if (_singleton != nullptr) {
+        AP_HAL::panic("Duplicate CRSF handler");
+    }
+
+    _singleton = this;
+#else
+    if (_singleton == nullptr) {
+        _singleton = this;
+    }
+#endif
+#if HAL_CRSF_TELEM_ENABLED && !APM_BUILD_TYPE(APM_BUILD_iofirmware) && !APM_BUILD_TYPE(APM_BUILD_UNKNOWN)
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_CRSF, 0);
+    if (_uart) {
+        start_uart();
+    }
+#endif
+}
+
+AP_RCProtocol_CRSF::~AP_RCProtocol_CRSF() {
+    _singleton = nullptr;
```

Comment:
Isn't this dead code?  If we never delete one, this is just wasted bytes.

Admittedly, not many wasted bytes.

Reply:
We delete one in RCProtocolTest - I had to do this to make it work